File: backend/F1Fantasy/Models/Race.cs
namespace F1Fantasy.Models;

public class Race
{
    public string Season { get; set; } = string.Empty;
    public string Round { get; set; } = string.Empty;
    public string Url { get; set; } = string.Empty;
    public string RaceName { get; set; } = string.Empty;
    public Circuit Circuit { get; set; } = new();
    public string Date { get; set; } = string.Empty;
    public string Time { get; set; } = string.Empty;
    public Session? FirstPractice { get; set; }
    public Session? SecondPractice { get; set; }
    public Session? ThirdPractice { get; set; }
    public Session? Qualifying { get; set; }
    public Session? Sprint { get; set; }
    public Session? SprintQualifying { get; set; }
}

public class Circuit
{
    public string CircuitId { get; set; } = string.Empty;
    public string Url { get; set; } = string.Empty;
    public string CircuitName { get; set; } = string.Empty;
    public Location Location { get; set; } = new();
}

public class Location
{
    public string Lat { get; set; } = string.Empty;
    public string Long { get; set; } = string.Empty;
    public string Locality { get; set; } = string.Empty;
    public string Country { get; set; } = string.Empty;
}

public class Session
{
    public string Date { get; set; } = string.Empty;
    public string Time { get; set; } = string.Empty;
}

// API Response models
public class ApiResponse
{
    public MRData? MRData { get; set; }
}

public class MRData
{
    public string? Xmlns { get; set; }
    public string? Series { get; set; }
    public string? Url { get; set; }
    public string? Limit { get; set; }
    public string? Offset { get; set; }
    public string? Total { get; set; }
    public RaceTable? RaceTable { get; set; }
}

public class RaceTable
{
    public string? Season { get; set; }
    public List<Race>? Races { get; set; }
}

File: backendJava/f1fantasy/src/main/java/no/f1fantasy/entity/Race.java
package no.f1fantasy.entity;

import jakarta.persistence.*;
import lombok.Data;

@Entity
@Table(name = "Races")
@IdClass(RaceId.class)
@Data
public class Race {

    @Id
    @Column(name = "Season", length = 10)
    private String season;

    @Id
    @Column(name = "Round", length = 10)
    private String round;

    @Column(name = "Url", length = 500)
    private String url;

    @Column(name = "RaceName", length = 200)
    private String raceName;

    @Column(name = "Date", length = 50)
    private String date;

    @Column(name = "Time", length = 50)
    private String time;

    // Circuit is NOT stored in the Races table (ignored by EF Core); returned from Ergast API only
    @Transient
    private Circuit circuit;

    @Embedded
    @AttributeOverrides({
            @AttributeOverride(name = "date", column = @Column(name = "FirstPractice_Date", length = 50)),
            @AttributeOverride(name = "time", column = @Column(name = "FirstPractice_Time", length = 50))
    })
    private Session firstPractice;

    @Embedded
    @AttributeOverrides({
            @AttributeOverride(name = "date", column = @Column(name = "SecondPractice_Date", length = 50)),
            @AttributeOverride(name = "time", column = @Column(name = "SecondPractice_Time", length = 50))
    })
    private Session secondPractice;

    @Embedded
    @AttributeOverrides({
            @AttributeOverride(name = "date", column = @Column(name = "ThirdPractice_Date", length = 50)),
            @AttributeOverride(name = "time", column = @Column(name = "ThirdPractice_Time", length = 50))
    })
    private Session thirdPractice;

    @Embedded
    @AttributeOverrides({
            @AttributeOverride(name = "date", column = @Column(name = "Qualifying_Date", length = 50)),
            @AttributeOverride(name = "time", column = @Column(name = "Qualifying_Time", length = 50))
    })
    private Session qualifying;

    @Embedded
    @AttributeOverrides({
            @AttributeOverride(name = "date", column = @Column(name = "Sprint_Date", length = 50)),
            @AttributeOverride(name = "time", column = @Column(name = "Sprint_Time", length = 50))
    })
    private Session sprint;

    @Embedded
    @AttributeOverrides({
            @AttributeOverride(name = "date", column = @Column(name = "SprintQualifying_Date", length = 50)),
            @AttributeOverride(name = "time", column = @Column(name = "SprintQualifying_Time", length = 50))
    })
    private Session sprintQualifying;
}
